tests/unit/imports/server/Model.ts

```diff
@@ -0,0 +1,392 @@
+      });
+      let model: Model<typeof schema>;
+      this.beforeAll(async function () {
+        model = await createTestModel(schema);
+      });
+
+      this.afterEach(function () {
+        resetClock();
+      });
+
+      it('populates only _id, createdAt, and lastWrittenAt on insert', async function () {
+        const id = await model.insertAsync({});
+        const record = (await model.findOneAsync(id))!;
+
+        assert.isOk(record);
+        assert.isString(record._id);
+        assert.instanceOf(record.createdAt, Date);
+        assert.instanceOf(record.lastWrittenAt, Date);
+        assert.isUndefined(record.updatedAt);
+      });
+
+      it('only updates updatedAt and lastWrittenAt on update', async function () {
+        const initialDate = new Date();
+        setClock(() => initialDate);
+
+        const id = await model.insertAsync({});
+        const record = (await model.findOneAsync(id))!;
+
+        assert.isOk(record);
+        assert.deepEqual(record.createdAt, initialDate);
+        assert.deepEqual(record.lastWrittenAt, initialDate);
+        assert.isUndefined(record.updatedAt);
+
+        const laterDate = new Date(initialDate.getTime() + 1000);
+        setClock(() => laterDate);
+
+        await model.updateAsync(id, {});
+        const updatedRecord = (await model.findOneAsync(id))!;
+
+        assert.isOk(updatedRecord);
+        assert.deepEqual(updatedRecord.createdAt, initialDate);
+        assert.deepEqual(updatedRecord.updatedAt, laterDate);
+        assert.deepEqual(updatedRecord.lastWrittenAt, laterDate);
+      });
+
+      it('populates createdAt on upsert, but not update', async function () {
+        const initialDate = new Date();
+        setClock(() => initialDate);
+
+        const id = Random.id();
+        await model.upsertAsync(id, {});
+        const record = (await model.findOneAsync(id))!;
+        assert.isOk(record);
+
+        assert.deepEqual(record.createdAt, initialDate);
+        assert.deepEqual(record.lastWrittenAt, initialDate);
+
+        const laterDate = new Date(initialDate.getTime() + 1000);
+        setClock(() => laterDate);
+
+        await model.upsertAsync(id, {});
+        const updatedRecord = (await model.findOneAsync(id))!;
+
+        assert.isOk(updatedRecord);
+        assert.deepEqual(updatedRecord.createdAt, initialDate);
+      });
+
+      it('does not populate updatedAt on upsert', async function () {
+        const initialDate = new Date();
+        setClock(() => initialDate);
+
+        const id = Random.id();
+        await model.upsertAsync(id, {});
+        const record = (await model.findOneAsync(id))!;
+        assert.isOk(record);
+
+        assert.isUndefined(record.updatedAt);
+
+        // It's a bit odd that updatedAt doesn't get set on an upsert that updates
+        // an existing document (rather than inserting a new one), but there's no
+        // $setOnNotInsert operator and this matches our historical behavior
```

Comment:
Huh, I didn't realize that was how `updatedAt` behaved today, but yeah, it looks like FeatureFlags never gets an `updatedAt` set.

Maybe it makes sense for collections we plan to purely upsert into to use `lastWriteTimestamp` instead of `updatedTimestamp`? (and maybe something with similar behavior for `updatedBy`?)

Reply:
When I realized that this was actually how `Base`'s `updatedAt` behaves today, my conclusion was mostly that we should just replace it with something more like `lastWriteTimestamp` across the board. I don't think there's a meaningful information difference between "updatedAt is unset" and "updatedAt has the same value as createdAt", and I don't think there's anywhere we need to (e.g.) query for "updatedAt is unset because this record has been inserted but never updated"

However, there's enough going on that I don't want to add unnecessary behavioral change on top of everything else, so I've made myself a note to follow up on this after this PR.

Reply:
I agree we should make such changes later/orthogonally, and agree with your general logic about what we should want here.

There's arguably a (probably small, kinda dumb) benefit at the margin of having less data to serialize/push down from the server/parse (dates are comparatively expensive fields), but we can presumably reap that benefit nearly as easily by being more intentional about field projection when publishing, especially with all the nice infrastructure from #1410.

Reply:
They are, for what it's worth, somewhat less expensive than they look. Chrome (at least) negotiates the `permessage-deflate` extension to websockets, and some very casual testing shows that we get pretty significant compression if the `updatedAt` timestamp is the same (or close to) the `createdAt` timestamp:

```javascript
import zlib from 'zlib';
const record = Models.FeatureFlags.findOne();

> record.updatedAt === undefined
true

> zlib.deflateSync(EJSON.stringify(record)).length
147
> zlib.deflateSync(EJSON.stringify(Object.assign({}, record, { updatedAt: record.createdAt }))).length
153
> zlib.deflateSync(EJSON.stringify(Object.assign({}, record, { updatedAt: new Date(record.createdAt.getTime() + 1) }))).length
154
> zlib.deflateSync(EJSON.stringify(Object.assign({}, record, { updatedAt: new Date(record.createdAt.getTime() + 2) }))).length
155
```

So as long as the differences in timestamps are relatively small (which we'd expect in the cases where we were introducing an `updatedAt` timestamp that previously wasn't present), we only see a 6-8 byte difference in length of the deflated string that actually gets sent over the wire (vs. roughly 36 bytes for an un-deflated `updateAt` field + value in EJSON encoding).